vesting/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode, HasCompact};
+use frame_support::{
+	decl_error, decl_event, decl_module, decl_storage, ensure,
+	traits::{Currency, ExistenceRequirement, LockIdentifier, LockableCurrency, WithdrawReasons},
+};
+use sp_std::{
+	cmp::{Eq, PartialEq},
+	vec::Vec,
+};
+// FIXME: `pallet/frame-` prefix should be used for all pallet modules, but currently `frame_system`
+// would cause compiling error in `decl_module!` and `construct_runtime!`
+// #3295 https://github.com/paritytech/substrate/issues/3295
+use frame_system::{self as system, ensure_signed};
+use sp_runtime::{
+	traits::{CheckedAdd, SimpleArithmetic, StaticLookup, Zero},
+	DispatchResult, RuntimeDebug,
+};
+
+mod mock;
+mod tests;
+
+/// The vesting schedule.
+///
+/// Benefits would be granted gradually, `per_period` amount every `period` of blocks
+/// after `start`.
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug)]
+pub struct VestingSchedule<BlockNumber, Balance: HasCompact> {
+	pub start: BlockNumber,
+	pub period: BlockNumber,
+	pub period_count: u32,
+	#[codec(compact)]
+	pub per_period: Balance,
+}
+
+impl<BlockNumber: SimpleArithmetic + Copy, Balance: SimpleArithmetic + Copy> VestingSchedule<BlockNumber, Balance> {
+	/// Returns the end of all periods, `None` if calculation overflows.
+	pub fn end(&self) -> Option<BlockNumber> {
+		self.period
+			.checked_mul(&self.period_count.into())?
+			.checked_add(&self.start)
+	}
+
+	/// Returns all locked amount, `None` if calculation overflows.
+	pub fn total_amount(&self) -> Option<Balance> {
+		self.per_period.checked_mul(&self.period_count.into())
+	}
+
+	/// Returns locked amount for a given `time`.
+	///
+	/// Note this func assumes `schedule.end()` calculation doesn't overflow, and it should be guaranteed by callers.
+	pub fn locked_amount(&self, time: BlockNumber) -> Balance {
+		(1..=self.period_count).fold(Zero::zero(), |acc, i| {
```

Comment:
can we do this without loop? something like `period_count.saturating_sub((time - start) / period) * per_period`

Reply:
Yes it's doable, but with tricky type conversions as `(time - start) / period` is `BlockNumber`, `period_count` is `u32`, and `per_period` is `Balance`.

Reply:
maybe we can make `period_count` to `BlockNumber` if it makes thing easier. `BlockNumber` is u32 anyway.

Reply:
By far defining `period_count` as `u32` makes calculations easier, since all calculations except this one are just between `u32` and `Balance/BlockNumber`, both of which have `From<u32>` bound.

I feel like the hard part of conversions is between `Balance` and `BlockNumber`, as the concrete types are determined at runtime, unless we introduce a conversion helper type. Making `period_count` as `BlockNumber` would introduce a lot of conversions between `Balance` and `BlockNumber`.

Reply:
I'll try to do the conversions and see if edge cases could be handled easily.